golem-worker-executor-base/src/durable_host/filesystem/types.rs

```diff
@@ -52,11 +52,12 @@ impl<Ctx: WorkerCtx> HostDescriptor for DurableWorkerCtx<Ctx> {
 
     fn write_via_stream(
         &mut self,
-        self_: Resource<Descriptor>,
+        fd: Resource<Descriptor>,
         offset: Filesize,
     ) -> Result<Resource<OutputStream>, FsError> {
+        self.fail_if_read_only(&fd)?;
         record_host_function_call("filesystem::types::descriptor", "write_via_stream");
-        HostDescriptor::write_via_stream(&mut self.as_wasi_view(), self_, offset)
+        HostDescriptor::write_via_stream(&mut self.as_wasi_view(), fd, offset)
```

Comment:
I played around a bit with wasmtime and I saw a few issues with preopened directories:

1. Preopened directories in the root don't work well. If I have a preopen for '/' and '/data' and list descriptors for '/' I don't see the descriptor for /data.
2. Preopen only works on the level of directories, meaning that we would have to make the whole dir readonly. This does not really work with the format that is specced for the files properties in the application manifest and is very confusing for users imo.
3. When playing around with complex setups I saw bad file descriptor errors a few times, but I'm not sure how to consistently trigger them.

Based on these I decided to implement this slightly differently.
1: Each worker maintains a shared cache of read only files /tmp/foo/{hash}. When a worker gets scheduled all required files are hard linked from the cache to the worker tmpdir. All read-write files are copied into the appropriate location.
2: The worker maintains a hashset of readonly paths that were loaded. All mutating operations in the HostDescriptor resolve the host path via the ResourceTable and check the hashset. If the path is present we fail with permission denied. This is very similar to how the preopen feature is [implemented](https://github.com/bytecodealliance/wasmtime/blob/af968461b0800325d61353e065bf5fe71abbefd4/crates/wasi/src/host/filesystem.rs#L4) in wasmtime.

Reply:
Maybe you can get rid of the need to do hard links as well by adding an extra path mapping in each filesystem host function. (Just an idea, not necessarily the way to go)

Reply:
This hope was that with the hard links and the permission overlay we can keep the implementation on our side as simple as possible. I also saw that they have very similar infrastructure in wasmtime, so maybe we can get them to expose an api that allows us to just give the file permission to them.